Sources/ServiceLifecycle/ServiceRunner.swift

```diff
@@ -0,0 +1,383 @@
+                    // when the user did not specify any shutdown signals. We just have to tolerate
+                    // this.
+                    continue
+
+                case nil:
+                    fatalError("Invalid result from group.next(). We checked if the group is empty before and still got nil")
+                }
+            }
+
+            return .success(())
+        }
+
+        self.logger.debug(
+            "Service lifecycle ended"
+        )
+        try result.get()
+    }
+
+    private func shutdownGracefully(
+        group: inout TaskGroup<ChildTaskResult>,
+        gracefulShutdownManagers: [GracefulShutdownManager]
+    ) async throws {
+        guard case .running = self.state else {
+            fatalError("Unexpected state")
+        }
+
+        // We have to shutdown the services in reverse. To do this
+        // we are going to signal each child task the graceful shutdown and then wait for
+        // its exit.
+        for (gracefulShutdownIndex, gracefulShutdownManager) in gracefulShutdownManagers.lazy.enumerated().reversed() {
+            self.logger.debug(
+                "Triggering graceful shutdown for service",
+                metadata: [
+                    self.configuration.logging.serviceKey: "\(self.services[gracefulShutdownIndex])",
+                ]
+            )
+
+            await gracefulShutdownManager.shutdownGracefully()
+
+            let result = await group.next()
+
+            switch result {
+            case .serviceFinished(let service, let index):
+                if index == gracefulShutdownIndex {
+                    // The service that we signalled graceful shutdown did exit/
+                    // We can continue to the next one.
+                    self.logger.debug(
+                        "Service finished",
+                        metadata: [
+                            self.configuration.logging.serviceKey: "\(service)",
+                        ]
+                    )
+                    continue
+                } else {
+                    // Another service exited unexpectedly
+                    self.logger.debug(
+                        "Service finished unexpectedly during graceful shutdown. Cancelling all other services now",
+                        metadata: [
+                            self.configuration.logging.serviceKey: "\(service)",
+                        ]
+                    )
+
+                    group.cancelAll()
+                    throw ServiceRunnerError.serviceFinishedUnexpectedly()
+                }
+
+            case .serviceThrew(let service, _, let error):
+                self.logger.debug(
+                    "Service threw error during graceful shutdown. Cancelling all other services now",
+                    metadata: [
+                        self.configuration.logging.serviceKey: "\(service)",
+                        self.configuration.logging.errorKey: "\(error)",
```

Comment:
general nit: not all Swift errors interpolate well, this is interesting to improve

Reply:
another general nit is that while i like the use of structured logging, need to make sure it actually outputs a useful logging statement in the default case of a "dumb" logging backend

Reply:
I agree that errors sometimes don't interpolate well but I don't think we can fix that here.

I also agree on the structured logging statement though I think this does output a useful logging statement. I rather prefer doing proper structured logging here than interpolating them into a string.